### DISCUSSION TITLE
Real time countdown implementation

### DISCUSSION THREAD

**Discussion by ysomad** [miguelgrinberg/Flask-SocketIO#1695] [Q&A] [answered]
Hello! Working on typing speed test app using `flask-socketio` and I have an issue with countdown implementation.

I have `keydown` event in which im checking is typing test started or no, storing `is_started` in session so on next `keydown` emits I have 2 cases:
1. Start typing test if `is_started` is not set (when typing test is started countdown should start and emit `countdown` event to client
2. Emit `keydown_reply` which is sending pressed key back to client

What I achieve: implemented countdown using Celery
Problems: 
1. Celery cannot use main socket io instance and because of that it's impossible to emit events to a specific socket (or i did not find any information how to do it)
2. On page refresh countdown still going and I need manually to revoke it

Question:
Is there any ideas how to implement countdown so it can be independent (and emitting events back to the client on background) from `keydown` event handler and so that it is possible to get his state in the same handler?

```python
# tasks.py

@celery.task
def start_countdown(seconds: int) -> bool:
    socketio = create_socketio(flask_app, False)
    while seconds > 0:
        socketio.emit('typing_test_countdown', seconds)
        seconds -= 1
        socketio.sleep(1)
    return True
```

```python
# events.py

@socketio.on('keydown')
def on_keydown(key: str):
    cfg = session.get('config')
    typing_test_finished = False

    if session.get('is_started') in (None, False):
        # Start typing test
        session['is_started'] = True
        result = start_countdown.delay(cfg.duration) # starts celery task with countdown
        typing_test_finished = result.get()

        emit('keydown_reply', {
            'key': key, 
            'correct': False, # not implemented
        })
    else:
        # Process typing test  
        emit('keydown_reply', {
            'key': key, 
            'correct': False, # not implemented
        })

    if typing_test_finished is True:
        # Send typing test results
        emit('get_result', {}) # not implemented
        typing_test_finished = False
        session['is_started'] = False
```

**Comment by miguelgrinberg** [comment, marked as answer]
My main question is why do you need the server to count down the seconds. Wouldn't it be easier for the client to count the time on its own? The server has to keep track of when the countdown expires, of course, but I don't see why it has to control the client side timer, since both can count seconds on their own and they'll be very close, if not exactly the same.

> Celery cannot use main socket io instance and because of that it's impossible to emit events to a specific socket (or i did not find any information how to do it)

See the section on emitting from an external process in the documentation.

> On page refresh countdown still going and I need manually to revoke it

You can use the disconnect handler to do any necessary clean up work. This is invoked after the user refreshes the page.

**Reply by ysomad** [reply]
That's true, I thought it would be more precisely for some reason but I'll stick with counting on client ofc, thank you!

I've done celery countdown issue but now I'm switching to `python-socketio` since I don't need Flask at all on this project.

Sorry for a long response, I was on sick leave.
